snuba/subscriptions/entity_subscription.py

```diff
@@ -0,0 +1,122 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Mapping, Optional, Sequence, Type
+
+from snuba.datasets.entities import EntityKey
+from snuba.query.conditions import ConditionFunctions, binary_condition
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.query.types import Condition
+
+
+class SubscriptionType(Enum):
+    LEGACY = "legacy"
+    SNQL = "snql"
+    DELEGATE = "delegate"
+
+
+class EntitySubscription(ABC):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        ...
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        legacy subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+
+class SessionsSubscription(EntitySubscription):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        super().__init__(data_dict)
+        try:
+            self.organization: int = data_dict["organization"]
+        except KeyError:
+            raise InvalidQueryException(
+                "organization param is required for any query over sessions entity"
+            )
+
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        return [
+            binary_condition(
+                ConditionFunctions.EQ,
+                Column(None, None, "org_id"),
+                Literal(None, self.organization),
+            ),
+        ]
+
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        return []
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {"organization": self.organization}
+
+
+class BaseEventsSubscription(EntitySubscription, ABC):
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
```

Comment:
```suggestion
    ) -> Sequence[Expression]:
```

Reply:
I don't think that's a good idea because you can't really use th `+` with `Sequence` here -> https://github.com/getsentry/snuba/blob/3ea9573691182f29541b15b8c42c558827bfe775/snuba/subscriptions/data.py#L246

Reply:
Hmm I see. How about:
```
conditions_to_add = [
  condition1,
  condiiton2,
  *self.entity_subscription.get_entity_subscription_conditions_for_snql(offset)
]
```
That is more similar to what you do above for the legacy subscription data version and I think there is no reason to give snql and legacy different types just because of using splat operator in one place and += in the other.

Reply:
Agreed